test_ffi/tests/test.js

```diff
@@ -417,6 +456,52 @@ call_stored_function();
 dylib.symbols.store_function_2(ptr(add10Callback));
 dylib.symbols.call_stored_function_2(20);
 
+function logManyParametersFast(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s) {
+  return symbols.log_many_parameters(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s);
+};
+
+%PrepareFunctionForOptimization(logManyParametersFast);
+logManyParametersFast(255, 65535, 4294967295, 4294967296, 123.456, 789.876, -1, -2, -3, -4, -1000, 1000, 12345.678910, 12345.678910, 12345.678910, 12345.678910, 12345.678910, 12345.678910, 12345.678910);
+%OptimizeFunctionOnNextCall(logManyParametersFast);
+logManyParametersFast(255, 65535, 4294967295, 4294967296, 123.456, 789.876, -1, -2, -3, -4, -1000, 1000, 12345.678910, 12345.678910, 12345.678910, 12345.678910, 12345.678910, 12345.678910, 12345.678910);
+assertOptimized(logManyParametersFast);
+
+function castU8U32Fast(x) { return symbols.cast_u8_u32(x); };
+
+%PrepareFunctionForOptimization(castU8U32Fast);
+console.log(castU8U32Fast(256));
+%OptimizeFunctionOnNextCall(castU8U32Fast);
+// Some ABIs rely on the convention to zero/sign-extend arguments by the caller to optimize the callee function.
+// If the trampoline did not zero/sign-extend arguments, this would return 256 instead of the expected 0 (in optimized builds)
+console.log(castU8U32Fast(256));
+assertOptimized(castU8U32Fast);
+
+function castU32U8Fast(x) { return symbols.cast_u32_u8(x); };
+
+%PrepareFunctionForOptimization(castU32U8Fast);
+console.log(castU32U8Fast(256));
+%OptimizeFunctionOnNextCall(castU32U8Fast);
+// Some ABIs rely on the convention to expect garbage in the bits beyond the size of the return value to optimize the callee function.
+// If the trampoline did not zero/sign-extend the return value, this would return 256 instead of the expected 0 (in optimized builds)
+console.log(castU32U8Fast(256));
+assertOptimized(castU32U8Fast);
+
+// TODO: this test currently fails in aarch64-apple-darwin (also in branch main!). The reason is v8 does not follow Apple's custom
+// ABI properly (aligns arguments to 8 byte boundaries instead of the natural alignment of the parameter type).
+// A decision needs to be taken:
+// 1. leave it broken and wait for v8 to fix the bug
+// 2. Adapt to v8 bug and follow its ABI instead of Apple's. When V8 fixes the implementation, we'll have to fix it here as well
+function addManyU16Fast(a, b, c, d, e, f, g, h, i, j, k, l, m) { return symbols.add_many_u16(a, b, c, d, e, f, g, h, i, j, k, l, m); };
+
+%PrepareFunctionForOptimization(addManyU16Fast);
+console.log(addManyU16Fast(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12));
+%OptimizeFunctionOnNextCall(addManyU16Fast);
+// Generally the trampoline tail-calls into the FFI function, but in certain cases (e.g. when returning 8 or 16 bit integers)
+// the tail call is not possible and a new stack frame must be created. We need enough parameters to have some on the stack
```

Comment:
Why is this? I don't quite get the limitation

Reply:
In particular, on platforms where the return value should have zeros in the upper bits, is a tail call being emit?

I understand that it shouldn't, for platforms that allow "garbage" data in the upper bits.

Reply:
Yes, a tail call is being emit on platforms where manual zeroing is not needed.

Reply:
> Yes, a tail call is being emit on platforms where manual zeroing is not needed.

Concretely, SysV and Win64 assume garbage in the higher bits of 8/16 bit return values, therefore they must be extended when casting them to 32 bits. In these cases the trampoline cannot tail-call. Aarch64-Apple enforces 8/16 bit return values to be extended, so the cast does not require an additional extension and the trampoline can always tail-call.

Reply:
Thank you both for the confirmation; consider this resolved.

Reply:
> and the trampoline can always tail-call.

FYI, this is no longer the case, with the introduction of u64 return values wrapped in Int32Arrays